src/lib.rs

```diff
@@ -0,0 +1,489 @@
+// derivative works of, display, perform, and distribute the Software and make,
+// use, sell, offer for sale, import, export, have made, and have sold the
+// Software and the Larger Work(s), and to sublicense the foregoing rights on
+// either these or other terms.
+//
+// This license is subject to the following condition: The above copyright
+// notice and either this complete permission notice or at a minimum a reference
+// to the UPL must be included in all copies or substantial portions of the
+// Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#[macro_use] extern crate lazy_static;
+#[macro_use] extern crate log;
+extern crate env_logger;
+
+use std::collections::HashMap;
+use std::sync::RwLock;
+use std::thread;
+use std::thread::JoinHandle;
+use std::cell::RefCell;
+
+
+/// Uniquely identifies a location in the user-program.
+///
+/// XXX Hard-coded to a simple program counter value.
+/// XXX This need to be user-specified, perhaps using a trait?
+type JitLocation = u32;
+
+/// The type used to count how many times a JitLocation has been interpreted.
+type HotCount = u32;
+
+/// The number of times a location is interpreted before we trace it.
+/// XXX make this configurable.
+const HOT_THRESHOLD: HotCount = 10;
+
+/// The shared trace store.
+///
+/// Maps a location to a trace (if one exists). Locations not in the map have not yet been traced.
+lazy_static! {
+    static ref TRACES: RwLock<HashMap<JitLocation, JitTrace>> = RwLock::new(HashMap::new());
+}
+
+/// Execute the trace for the specified location (if there is one).
+///
+/// Returns true if a trace was present (and thus was executed), or false otherwise.
+fn try_exec_trace(loc: JitLocation) -> bool {
+    let trace_map = TRACES.read().unwrap();
+    let trace = trace_map.get(&loc);
+    if trace == Some(&JitTrace::FinalisedTrace) {
+        JitThreadContext::phase_transition(JitPhase::RunningTrace(loc));
+        // XXX Really execute the trace.
+        JitThreadContext::phase_transition(JitPhase::Interpreting);
+        return true;
+    }
+    false
+}
+
+/// Is a thread currently tracing/compiling the specified location?
+fn trace_pending(loc: JitLocation) -> bool {
+    let trace_map = TRACES.read().unwrap();
+    match trace_map.get(&loc) {
+        Some(&JitTrace::TracePending) => return true,
+        _ => return false,
+    };
+}
+
+/// Store a trace (or "trace pending" marker in the trace cache for the specified location.
+fn set_trace(loc: JitLocation, trace: JitTrace) {
+    // Sanity checks.
```

Comment:
I think these sanity checks might be wrong. I think two threads can trigger a trace at the same point, before the lock is gained on line 98. If I'm right, then what we should do is attempt to notice if another thread is also tracing, and, if so, stop doing so oureslves.

Reply:
I think the correct fix is proactive, nit reactive. Ensure that not two threads are tracing at the same time.

There is another race in this function, again between the two locks.